doc/adr/0004-molgenis-core-apps.md

```diff
@@ -0,0 +1,55 @@
+# 4. Molgenis Core Apps
+
+Date: 2020-09-23
+
+## Status
+
+Proposal
+
+## Context
+
+See [frontend architecture sessions](https://docs.google.com/document/d/1VW3ah5VAvAz2KnqNZlNmVqCzFhBMlIcjPPUlsHMFRIY/) for additional background
+information on micro-frontend/SPA architecture.
+
+Molgenis is flexible software that allows users to create their own apps.
+This is quite powerful whenever a new, custom application needs to be made
+on top of the Molgenis backend. However, the same modular paradigm is also
+being applied to parts of the frontend, that are always being shipped with
+Molgenis.
+
+At this time, we have approximately 12 legacy Molgenis apps that are still
+rooted in Molgenis freemarker templates and Java controllers. There are
+about 11 new-style Unpkg frontend applications. The Unpkg applications
+all come with their own package configuration, release cycle, build tools
+and testsuites; all glued together in the Nginx proxy, so they are
+approachable from the same domain.
+
+## Decision
+
+The keyword here is convergence; the act of converging and especially moving
+toward union or uniformity. This means that Molgenis frontend core is going
+to be one Vue application, in which all components from the 11 existing Unpkg
```

Comment:
The big question(s) is/are: Do we need to be able to update the parts seperately and are some of the features 'opt-in' ?

Reply:
Would you consider using a multi-page setup to combine the apps together, instead of one big merge?

Reply:
That would be an interesting option if we decide to merge all buildtools from all packages together, while still keeping multiple apps around. However, I think we could make better use of Vue-router and Vuex by having only one application with multiple views. The main reasons:
* We could apply history routing to all parts of the application, making the application as a whole feel much more responsive.
* Easier dependency management, since we only have to include dependencies once. While we keep tests and sandbox environments separate, we still have to include all dependencies for each package(as dev-dependencies, but still...)

We could start a migration path by implementing just one simple app in a new package (e.g. core-app); for example a theme selector. The time we spend there would be useful to decide on definitive tooling. After this, we could migrate an existing package into the core-app; for example DE2. 

Right now, the current build toolchain (Webpack/Babel) is a bottleneck. As soon we move in more components in the build, we would start to see unacceptable buildtimes during development or while building for production(less urgent). More recent toolchains like Vite and Snowpack rely on ES modules during development and a high performance bundler (esbuild). That would definitly fix slow build times during development.

Webpack could use esbuild, but tweaking Webpack's config is complex, even more with the Vue-CLI wrapper around it. Vite(https://github.com/vitejs/vite#how-and-why) seems an interesting alternative. It uses the same kind of config as vue-cli, but without Webpack. Sadly, it has no Vue 2 support and Bootstrap-vue doesn't have a Vue 3 version yet(https://github.com/bootstrap-vue/bootstrap-vue/issues/5196). So that limits our Vite options to either waiting for a Vue3 version of Bootstrap-Vue and stick with Vue2 for now, or dropping Bootstrap-Vue and migrate each app to Vue 3.

Another option would be to use Snowpack together with Vue 2. I've tried Snowpack a while and it works really well. However, I think we should definitely give Vite a try first, since it treats Vue as a first-class citizen.